### PR TITLE
Sync

### PR DESCRIPTION
**Pull request by DieGurke**
* Added new Interface, which only uses one web request
* Added message state synchronization
* Added user status synchronization

### CODE REVIEW

File: src/main/java/envoy/client/ui/ChatWindow.java [hunk 1]
```diff
@@ -200,16 +199,23 @@ public ChatWindow(Client client, LocalDB localDB) {
...
 
 				textPane.setText(currentChat.getRecipient().getName());
+				
 				messageList.setModel(currentChat.getModel());
 				contentPane.revalidate();
 			}
 		});
-
+		
+
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/ChatWindow.java]: Too many empty lines.

File: src/main/java/envoy/client/ui/ChatWindow.java [hunk 2]
```diff
@@ -253,18 +260,32 @@ private void loadUsersAndChats() {
 	}
 
 	/**
-	 * Checks for new messages and adds them to the chat list.
+	 * For detailed information see Javadoc of corresponding methods.
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/ChatWindow.java]: Describe the method, add references with the `@see` tag.

File: src/main/java/envoy/client/ui/ChatWindow.java [hunk 3]
```diff
@@ -147,18 +147,17 @@ public ChatWindow(Client client, LocalDB localDB) {
...
 
 			if (!messageEnterTextfield.getText().isEmpty()) try {
 
 				// Create and send message object
 				final Message message = client.createMessage(messageEnterTextfield.getText());
-				client.sendMessage(message);
-
-				// Append message object to chat
-				currentChat.appendMessage(message);
+				client.addWaitingMessageToLocalDB(message, currentChat);
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/ui/ChatWindow.java]: Add the pending message directly to LocalDB and remove the dependency from Client.

File: src/main/java/envoy/client/Client.java [hunk 4]
```diff
@@ -46,129 +47,416 @@ public Client(Config config, String username) {
...
-	 * @param recipient   name of the recipient
-	 * @param textContent content (text) of the message
+	 * @param textContent The content (text) of the message
+	 * @return prepared {@link Message} object
 	 */
-	public void sendMessage(Message message) {
-		new Thread(() -> {
-			// Wrap single message into messages list
-			Messages messages = wrapMessages(message);
+	public Message createMessage(String textContent) {
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/Client.java]: Move to LocalDB.

File: src/main/java/envoy/client/Client.java [hunk 5]
```diff
@@ -46,129 +47,416 @@ public Client(Config config, String username) {
...
+	 */
+	public Sync getUsersListXml() {
+		Sync	sendSync	= objectFactory.createSync();
+		User	user		= objectFactory.createUser();
+		user.setID(-1);
+		sendSync.getUsers().add(user);
+
+		javax.ws.rs.client.Client	client	= ClientBuilder.newClient();
+		WebTarget					target	= client.target(String
+			.format("%s:%d/envoy-server/rest/sync/syncData?userId=%d", config.getServer(), config.getPort(), 0));
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/Client.java]: Extract to generic `get `method.

File: src/main/java/envoy/client/Client.java [hunk 6]
```diff
@@ -46,129 +47,416 @@ public Client(Config config, String username) {
...
+	private User getUser(String name) {
+		Sync	senderSync	= objectFactory.createSync();
+		User	user		= objectFactory.createUser();
+		user.setName(name);
+		senderSync.getUsers().add(user);
+
+		javax.ws.rs.client.Client	client	= ClientBuilder.newClient();
+		WebTarget					target	= client.target(String
+			.format("%s:%d/envoy-server/rest/sync/syncData?userId=%d", config.getServer(), config.getPort(), 0));
+
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/Client.java]: Extract to generic `post` method.

File: src/main/java/envoy/client/Client.java [hunk 7]
```diff
@@ -46,129 +47,416 @@ public Client(Config config, String username) {
...
+	 * Users: <br>
+	 * Updating UserStatus of all users in LocalDB. (Server sends all users with their updated UserStatus to the client.) <br>						
+	 * 
+	 * @param userId
+	 * @param localDB
+	 * @since Envoy v0.1-alpha
+	 */
+	public void sendSync(long userId, LocalDB localDB) {
+		new Thread(() -> {
+			// Print sync XML to console
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/Client.java]: Extract to `printXML` method, possibly in a utility class.

File: src/main/java/envoy/client/Client.java [hunk 8]
```diff
@@ -46,129 +47,416 @@ public Client(Config config, String username) {
...
+	 * 									the message(s).) <br>
+	 * <br>
+	 * Users: <br>
+	 * Updating UserStatus of all users in LocalDB. (Server sends all users with their updated UserStatus to the client.) <br>						
+	 * 
+	 * @param userId
+	 * @param localDB
+	 * @since Envoy v0.1-alpha
+	 */
+	public void sendSync(long userId, LocalDB localDB) {
```
Comment (CyB3RC0nN0R) [inline src/main/java/envoy/client/Client.java]: Move this method into `LocalDB` and split it into two parts:

- [ ] `applySync` extracts changes from a sync object and updates the corresponding data structures in `LocalDB`. Call this method from `Client`.

- [ ] Process changes of synchronized objects in `LocalDB` to reduce dependencies between `Sync` and other classes.

File: src/main/java/envoy/client/Client.java [hunk 9]
```diff
@@ -110,321 +86,74 @@ private User getUser(String name) {
...
 
 	/**
 	 * Sends the "sync" Sync Object to the server and gets a "returnSync" Sync
 	 * Object as response. <br>
-	 * It is also used to get the own sender at the start of the client 
-	 * (Client sends "sync" Sync Object with single user in it(name: the name entered at login, id: 0, UserStatus:null))<br>
-	 *  and to get a complete list of all users saved on the server. 
-	 * (Client sends "sync" Sync Object with single user in it(name: "" (empty), id: -1, UserStatus:null)) <br>
+	 * It is also used to get the own sender at the start of the client
+	 * (Client sends "sync" Sync Object with single user in it(name: the name
```
Comment (delvh) [inline src/main/java/envoy/client/Client.java]: please change Sync to `{@code Sync}` or `{@linc Snyc}`

File: src/main/java/envoy/client/Client.java [hunk 10]
```diff
@@ -110,321 +86,74 @@ private User getUser(String name) {
...
+	 * the message(s).) <br>
 	 * <br>
 	 * Users: <br>
-	 * Updating UserStatus of all users in LocalDB. (Server sends all users with their updated UserStatus to the client.) <br>						
+	 * Updating UserStatus of all users in LocalDB. (Server sends all users with
+	 * their updated UserStatus to the client.) <br>
 	 * 
 	 * @param userId
-	 * @param localDB
 	 * @since Envoy v0.1-alpha
```
Comment (delvh) [inline src/main/java/envoy/client/Client.java]: - [ ] please add `@param`, @DieGurke

File: src/main/java/envoy/client/LocalDB.java [hunk 11]
```diff
@@ -87,6 +107,276 @@ private void loadFromLocalDB() throws EnvoyException {
...
+			addWaitingMessagesToSync();
+
+			getSentStateMessagesFromLocalDB();
+			for (int i = 0; i < readMessages.getMessages().size(); i++) {
+				sync.getMessages().add(readMessages.getMessages().get(i));
+			}
+			readMessages.getMessages().clear();
+			
+
+		System.out.println(sync.getMessages().size());
```
Comment (delvh) [inline src/main/java/envoy/client/LocalDB.java]: - [ ] Please add that either to the log, comment it out or delete it.

File: src/main/java/envoy/client/LocalDB.java [hunk 12]
```diff
@@ -87,6 +107,276 @@ private void loadFromLocalDB() throws EnvoyException {
...
+					.get(i)
+					.getMetadata()
+					.getSender()) {
+					getChats().get(j).appendMessage(unreadMessages.getMessages().get(i));
+				}
+	}
+
+	/**
+	 * Gets all Messages with state SENT from the LocalDB and adds them to the
+	 * "sync" Sync object.
```
Comment (delvh) [inline src/main/java/envoy/client/LocalDB.java]: - [ ] please make `Sync ` to `{@link / @code Sync}`

File: src/main/java/envoy/client/LocalDB.java [hunk 13]
```diff
@@ -87,6 +107,276 @@ private void loadFromLocalDB() throws EnvoyException {
...
+			}
+		}
+	}
+
+	/**
+	 * Changes all Messages with State RECEIVED of a specific chat to State READ.
+	 * <br>
+	 * Adds these Messages to the "readMessages" Sync object.
+	 * 
+	 * @param currentChat
```
Comment (delvh) [inline src/main/java/envoy/client/LocalDB.java]: - [ ] Please add further param description

File: src/main/java/envoy/client/LocalDB.java [hunk 14]
```diff
@@ -87,6 +107,276 @@ private void loadFromLocalDB() throws EnvoyException {
...
+					readMessages.getMessages().add(currentChat.getModel().get(j));
+				}
+			}
+		}
+	}
+
+	/**
+	 * Adds a Message with State WAITING to a specific chat in the LocalDB.
+	 * 
+	 * @param message
```
Comment (delvh) [inline src/main/java/envoy/client/LocalDB.java]: - [ ] Please add further param description

File: src/main/java/envoy/client/ui/ChatWindow.java [hunk 15]
```diff
@@ -266,20 +263,23 @@ private void loadUsersAndChats() {
 	 *                the server
```
Comment (delvh) [inline src/main/java/envoy/client/ui/ChatWindow.java]: - [ ] Please elaborate on Javadoc, @DieGurke

**Pull request merged**